### ISSUE TITLE
Support for weighted-edges PageRank

### ISSUE THREAD

**Issue by felixcheung** [graphframes/graphframes#148]
There has been several requests and having running into the need for this myself, I want to see if it would be useful to add support for weighted edges for PageRank.

This would include a fairly contained change to add computing weight from edges with numeric weight values, in addition to the current implementation of computing weight from outDegrees. I would propose to add a `useEdgeWeight` (for an implicitly named `weight` attribute column on the edges) or `weightCol` (similar to spark.ml) and we could have this first implemented in backport/PageRank.scala (yes this would mean we would be forking it from the Spark GraphX implementation for the time being)

Would this be of interests?
If it is I would be happy to work on this.

**Comment by felixcheung** [comment]
@thunterdb @jkbradley @mengxr

**Comment by jkbradley** [comment]
I agree this would be a good thing to add, but I hesitate to fork further from GraphX.  What would you think about implementing it on DataFrames?  The new Connected Components implementation has worked out so far from what I've heard, so it might be cool to try another algorithm.

**Comment by hugolouzada** [comment]
@felixcheung that would be rather useful. Thanks for your offer to work on this.

If I may add one suggestion, I vote something like the weightCol option, but instead a parameter that could accept a list of attribute columns to act as weight. Then, the implementation could allow the user to run several weighted PageRanks in one go. A feature like this would definitely be useful for the data exploration perspective, as we often don't know which weight to use a priori.

**Comment by felixcheung** [comment]
@jkbradley great point, that sounds like a better approach though it might be a bigger project to take on. Would there be any existing benchmark & results to compare against?

re @hugolouzada comment, what do we think about the API to support both (existing) outDegrees and (new) weighted edges - or multiple weights?

**Comment by jkbradley** [comment]
I don't know if there are standard benchmarks, but there are certainly lots of big graph datasets which would be good candidates for PageRank.  Generating extreme graphs can also be useful: for Connected Components, we did a lot of testing with grids, stars, chains, and combinations of these.  The main things we tested were speed vs. the GraphX implementation for small/medium graphs and scalability to larger graphs than GraphX could handle.  If you'd like to work on this, then definitely check out the new CC implementation since it has to do some hacks like checkpointing.

@hugolouzada Could you please say more about the use case requiring multiple edge weights?  I feel like that's less common since edge weights are often computed based upon some pre-determined function (say something measuring similarity between Wikipedia articles, if each graph vertex is an article).  It would be helpful to know more.

**Comment by hugolouzada** [comment]
@felixcheung supporting both sounds about right, though I'm not sure what would be the implications for the users. I'm not that familiar with the implementation to have a better opinion, sorry.

@jkbradley I was thinking about data exploration. I believe that if you already have a good idea of what insights you might get from PageRank, than usually the choice of weight comes naturally. The problem would be when you want to use the rank value as something for later decisions. One might be in a situation with several weights, not knowing a priori which one will actually "work" to predict some feature. Take an air transportation graph as example, where vertices are airports and edges are connecting flights. One could weight it by the number of flights between airports, or the number of passengers transported, or the sum of cargo value. In all cases the weights might have some correlation, but each Weighted PageRank will show us different airports. Later, if you wanna know which airports to make a priority for some policy actions, like government funding, you might need to consider a combination of all three PageRanks to reorder the list.

**Comment by aray** [comment]
It would make sense to have an optional parameter for a user supplied weight `Column` with default value `null` that triggers the existing behavior.

I'll also throw my hat in the ring and say I would be willing to reimplement pagerank with dataframes.

**Comment by felixcheung** [comment]
Thanks Andrew.

I'm already making a lot of head way on implementing PageRank.

**Comment by thunterdb** [comment]
@felixcheung that sounds great!

**Comment by makalaaneesh** [comment]
@felixcheung Hey, I've come across this exact need on my current project. Great idea! Just wanted to know the status on this feature. Also, GraphFrames hasn't yet interfaced this GraphOps function [staticParallelPersonalizedPageRank](https://spark.apache.org/docs/2.1.0/api/java/org/apache/spark/graphx/GraphOps.html#staticParallelPersonalizedPageRank(long[],%20int,%20double)) (Introduced in Graphx 2.1.0). Is this going to be supported or is the plan to reimplement it on Dataframes still on? Thanks!

**Comment by jkbradley** [comment]
@makalaaneesh  It'd be nice to wrap personalize pagerank for GraphFrames.  Would you mind creating a new issue to track that?

**Comment by sui414** [comment]
@felixcheung it would be really helpful if it could have weighted edge version as it would be universally applicable to all DAG network analysis

**Comment by felixcheung** [comment]
sorry - I completely had to drop because of work. Hopefully we could get back on this...

**Comment by mohit-udaan** [comment]
> sorry - I completely had to drop because of work. Hopefully we could get back on this...

(Reviving this request) @felixcheung @aray Any plans for picking this up. Would be a great/practical feature.

**Comment by SemyonSinchenko** [comment]
Weights handling should be done at the level of the graph, not per algorithm. Closed in favor of #565

**Comment by rjurney** [comment]
In this instance there is a very easy fix. The PageRank implementation already uses weights it computes. We merely need to allow a preexisting column named `weight` to be used and normalized to sum to 1 at the start of the algorithm.

**Comment by SemyonSinchenko** [comment]
> In this instance there is a very easy fix. The PageRank implementation already uses weights it computes internally. We merely need to allow a preexisting column named `weight` to be used and normalized to sum to 1 at the start of the algorithm.


How to name the column? What data type should it be? Should we check the data type? How to handle nulls in this column?

While the fix is easy, it requires changes in API... And after these changes are here, deprecation or changing them will be a very long story because we should follow a sem-ver.

That is the main reason I would like to introduce graph properties and attributes (weights, direction, etc.) BEFORE propogating these changes to algorithms...

Otherwise we are risking to have a mess instead of clear API.

**Comment by rjurney** [comment]
The column has a name already, the one everyone uses and so should we: `weight`. It is a double. Impute nulls to zeroes or die happily. I’m not in disagreement about a PropertyGraphFrame class.
